### ISSUE TITLE
SIMD EqualsAll and Equals generate inefficient code

### ISSUE THREAD

**Issue by jbayardo** [dotnet/coreclr#8181]
We have the following snippet of code (and corresponding generated assembler):

```ASM
                var originalData = Unsafe.Read<Vector<byte>>(Original + i);
00007FFDB1BF8EC7  mov         r8,rdi  
00007FFDB1BF8ECA  add         r8,qword ptr [rsi+10h]  
00007FFDB1BF8ECE  movups      xmm0,xmmword ptr [r8]  
                var modifiedData = Unsafe.Read<Vector<byte>>(Modified + i);
00007FFDB1BF8ED2  mov         r8,rdi  
00007FFDB1BF8ED5  add         r8,qword ptr [rsi+18h]  
00007FFDB1BF8ED9  movups      xmm1,xmmword ptr [r8]  

                _allZeros &= modifiedData == Vector<byte>.Zero;
00007FFDB1BF8EDD  pxor        xmm2,xmm2  
00007FFDB1BF8EE1  movaps      xmm3,xmm1  
00007FFDB1BF8EE4  pcmpeqd     xmm3,xmm2  
00007FFDB1BF8EE8  pshufd      xmm4,xmm3,4Eh  
00007FFDB1BF8EED  andps       xmm3,xmm4  
00007FFDB1BF8EF0  pshufd      xmm4,xmm3,1  
00007FFDB1BF8EF5  pand        xmm3,xmm4  
00007FFDB1BF8EF9  movd        r8d,xmm3  
00007FFDB1BF8EFE  cmp         r8d,0FFFFFFFFh  
00007FFDB1BF8F05  sete        r8b  
00007FFDB1BF8F09  movzx       r8d,r8b  
00007FFDB1BF8F0D  and         byte ptr [rsi+31h],r8b  
```

EqualsAll generates similar code for `Vector.EqualsAll(modifiedData, Vector<byte>.Zero)`. Shouldn't the same result be achievable through xor and ptest (sse4.1), or even phadd (ssse3.1)?. Also, the loading could be done using `movdqa`/`movdqu` instead of `movaps` and avoid the floating point conversion altogether.

**Comment by redknightlois** [comment]
This is related to ```System.Numerics``` being designed over the idea of floating point calculations. Relates to https://github.com/dotnet/coreclr/issues/6906

cc @mellinoe @terrajobst

**Comment by mellinoe** [comment]
> being designed over the idea of floating point calculations

It's not; `Vector<T>` supports all of the primitive numeric types. Could you explain what you mean by that?

**Comment by redknightlois** [comment]
That it makes sense that loads being done as floating point because of the usual use case for `System.Numerics` and from there, most codegen takes that assumption over. 

If that is not the case, why a `Vector<byte>` gets loaded with a floating point operation is beyond my understanding. Reciprocal throughput and latency looks like to be better on `movdqa` than on `movaps` for most of the usual architectures out there. It looks more like a left-over of the floating point assumption than anything else. After all the code works anyways as it is.

**Comment by mikedn** [comment]
> That it makes sense that loads being done as floating point because of the usual use case for System.Numerics and from there, most codegen takes that assumption over. 

Weird assumption.

> If that is not the case, why a Vector<byte> gets loaded with a floating point operation is beyond my understanding.

`movups` is one byte shorter.

> Reciprocal throughput and latency looks like to be better on movdqa than on movaps for most of the usual architectures out there. 

Depends on what you mean by "usual". Is, say, Haswell usual? Then they have exactly the same latency, throughput and port usage.

> Also, the loading could be done using movdqa/movdqu instead of movaps and avoid the floating point conversion altogether.

What floating point conversions?!?

**Comment by redknightlois** [comment]
@mikedn That makes much more sense, being `movups` a byte shorter is a net win there. Read wrongly the latency tables (dont understand where I mixed up, now I copied into notepad and yes they are equal). Having said that, even accounting for that small difference I err on, wouldnt offset being a byte shorter anyways.

**Comment by terrajobst** [comment]
Looping in @CarolEidt

**Comment by sivarv** [comment]
@jbayardo - Right now RyuJIT supports SSE2 and AVX2 instruction set for SIMD codegen.  With PR https://github.com/dotnet/coreclr/pull/7367  I have fixed `Equals` against `Vector<T>.Zero `to use ptest.  since ptest is available in SSE4.1 and above, you will see the benefit when running on an AVX2 machine.

Regardings `Vector.EqualsAll(modifiedData, Vector<byte>.Zero)`, Can you paste what is the code that is getting generated?  As per my understanding it should generate `PCMPEQB` for comparison.

**Comment by jbayardo** [comment]
@sivarv The code being generated when using EqualsAll:

```ASM
                _allZeros &= Vector.EqualsAll(modifiedData, Vector<byte>.Zero);
00007FFDB1BC935D  movzx       r8d,byte ptr [rsi+31h]  
00007FFDB1BC9362  movaps      xmm2,xmm1  
00007FFDB1BC9365  pxor        xmm3,xmm3  
00007FFDB1BC9369  movaps      xmm4,xmm2  
00007FFDB1BC936C  pcmpeqd     xmm4,xmm3  
00007FFDB1BC9370  pshufd      xmm5,xmm4,4Eh  
00007FFDB1BC9375  andps       xmm4,xmm5  
00007FFDB1BC9378  pshufd      xmm5,xmm4,1  
00007FFDB1BC937D  pand        xmm4,xmm5  
00007FFDB1BC9381  movd        ecx,xmm4  
00007FFDB1BC9385  cmp         ecx,0FFFFFFFFh  
00007FFDB1BC938B  sete        cl  
00007FFDB1BC938E  movzx       ecx,cl  
00007FFDB1BC9391  and         ecx,0FFh  
00007FFDB1BC9397  and         ecx,r8d  
00007FFDB1BC939A  mov         byte ptr [rsi+31h],cl  
```

The same issue appears when using == between non-necessarily-zero vectors:

```ASM
                if (modifiedData == originalData)
00007FFDB1BC939D  movaps      xmm2,xmm1  
00007FFDB1BC93A0  pcmpeqd     xmm2,xmm0  
00007FFDB1BC93A4  pshufd      xmm3,xmm2,4Eh  
00007FFDB1BC93A9  andps       xmm2,xmm3  
00007FFDB1BC93AC  pshufd      xmm3,xmm2,1  
00007FFDB1BC93B1  pand        xmm2,xmm3  
00007FFDB1BC93B5  movd        r8d,xmm2  
00007FFDB1BC93BA  cmp         r8d,0FFFFFFFFh  
00007FFDB1BC93C1  sete        r8b  
00007FFDB1BC93C5  movzx       r8d,r8b  
00007FFDB1BC93C9  test        r8d,r8d  
00007FFDB1BC93CC  je          00007FFDB1BC93F4  
```

I have been looking through the Intel manual and have a few ideas, maybe they'll save you some time:

- Load up the 128 bits altogether in memory (`movdqa` or `movdqu` are both supported in SSE2)
- `pxor` (also sse2) supports `xmm1, xmm2/m128`, so strictly speaking `pxor xmm1, [memory location]` could be done if one of the values is not in the registers and not required later, which would save one `movdq(a/u)`.

In terms of the actual comparison, it seems to me that the best strategy depends on the SSE version available:

- 4.1/4.2: `ptest`
- ssse3: `phaddsw` maybe?
- sse2: `cmppd` with zero and then `pshufd`

It seems to me that #6906 would be indeed a better way to solve the issue though. Nevertheless, it would be nice to see this fixed either way, as these sort of issues have forced us to drop the idea of using SIMD twice already.

**Comment by mikedn** [comment]
> Load up the 128 bits altogether in memory (movdqa or movdqu are both supported in SSE2)

As already mentioned this isn't very useful.

> pxor (also sse2) supports xmm1, xmm2/m128, so strictly speaking pxor xmm1, [memory location] could be done if one of the values is not in the registers and not required later, which would save one movdq(a/u).

To be precise that saves a `movdqa`, not a `movdqu`. That makes this useless in this particular case.

> sse2: cmppd with zero and then pshufd 

Floating point operations on non-floating point data?!?

> The same issue appears when using == between non-necessarily-zero vectors:

> Regarding Vector.EqualsAll(modifiedData, Vector<byte>.Zero), Can you paste the code that is getting generated? As per my understanding it should generate PCMPEQB for comparison.

There seems to be a confusion between `Equals` and `EqualsAll` throughout this issue. `==`, `bool Vector<T>.Equals(Vector<T> v)` and `bool Vector.EqualsAll<T>(Vector<T> v1, Vector<T> v2)` are exactly the same thing. They compare 2 vectors for equality and return a `bool`, that makes their implementation potentially problematic depending on the available instructions.

There's also `Vector<T> Vector.Equals(Vector<T> v1, Vector<T> v2)`. This one generates a single `PCMPEQB` as expected.

In fact, it looks like you should have used this version of Equals for the allZeroes thing inside the for loop. Going from `Vector<byte>` to `bool` should have been done outside the loop.

**Comment by jbayardo** [comment]
> Floating point operations on non-floating point data?!?

I'm sorry, I missed on that, it should have been `pcmpeqd` and `pshufd`.

> There seems to be a confusion between Equals and EqualsAll throughout this issue. ==, bool Vector<T>.Equals(Vector<T> v) and bool Vector.EqualsAll<T>(Vector<T> v1, Vector<T> v2) are exactly the same thing. They compare 2 vectors for equality and return a bool, that makes their implementation potentially problematic depending on the available instructions.

The `bool` returning equality functions could be implemented as mentioned, while the `Vector<int/byte/word>` returning ones could be just pcmpeq(d|b|w) accordingly. As-is, the `bool` returning functions are implemented in terms of pcmpeq, which ends up taking more instructions.

We have been looking into a possible sse2 version, and came up with this:

```asm
// xmm1 = p1, xmm2 = p2

PXOR 	xmm1, xmm2    // if any bit is different then it is going to be 1
MOVQ 	rax, xmm1     // move the lower part to rax
PSRLDQ 	xmm1, 64      // shift by half
MOVQ 	rbx, xmm1     // move the lower part to rbx
OR 	rax, rbx      // 1 if there is any bit different. 
JNZ/JZ		      // The actual jump for the if condition, if its there
```

We haven't tested it. Latency count seems to be better, as we have restricted the XMM operations and distributed the cost over the functional units. We are nevertheless worried about possible pipeline stalls. This may be a faster alternative to the `bool`-returning functions.

**Comment by mikedn** [comment]
> We have been looking into a possible sse2 version, and came up with this:

This might be a bit better:
```
pcmpeqb     xmm0, xmm1
pmovmskb    ecx, xmm0
cmp         ecx, 0x0FFFF
```

**Comment by mikedn** [comment]
@sivarv Is there any technical reason why the JIT doesn't use `pmovmskb` for compares? I modified the JIT to emit `pmovmskb` and it seems to work fine but maybe I'm missing something.

**Comment by sivarv** [comment]
@mikedn - I agree equality compare that produces bool result can be improved using pmovmskb.  I was thinking of trying it out myself.  In case if you have a PR ready,  pleasse go ahead.  Otherwise I can take over this issue.

**Comment by mikedn** [comment]
@sivarv OK then, I'll finish my changes and send a PR.

**Comment by redknightlois** [comment]
Is this going to be released on a service release for 1.1 or we wanna wait until 2.0?

**Comment by sivarv** [comment]
This will go in 2.0
